components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt

```diff
@@ -0,0 +1,143 @@
+ */
+class DownloadManager(private val applicationContext: Context) {
+
+    private val queuedDownloads = HashMap<Long, OnDownloadFinish>()
+    private var isSubscribedReceiver = false
+    private lateinit var androidDownloadManager: AndroidDownloadManager
+
+    /**
+     * Schedule a download through the [AndroidDownloadManager].
+     * @param download metadata related to the download.
+     * @param downloadListener a listener to be notify when the download is completed.
+     * @param refererURL the url from where this download was referred.
+     * @param cookie any additional cookie to add as part of the download request.
+     * @return the id reference of the scheduled download.
+     */
+    @RequiresPermission(allOf = arrayOf(INTERNET, WRITE_EXTERNAL_STORAGE))
+    fun download(
+        download: Download,
+        downloadListener: OnDownloadFinish = {},
+        refererURL: String = "",
+        cookie: String = ""
+    ): Long {
+
+        if (!applicationContext.isPermissionGranted(INTERNET, WRITE_EXTERNAL_STORAGE))
+            throw SecurityException("You must be granted INTERNET and WRITE_EXTERNAL_STORAGE permissions")
+
+        androidDownloadManager = applicationContext.getSystemService(DOWNLOAD_SERVICE) as AndroidDownloadManager
+
+        val fileName = getFileName(download)
+
+        val request = Request(Uri.parse(download.url))
+            .setNotificationVisibility(VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
+            .setMimeType(download.contentType)
+            .addRequestHeader("User-Agent", download.userAgent)
+
+        if (cookie.isNotEmpty()) {
+            request.addRequestHeader("Cookie", cookie)
+        }
+
+        if (refererURL.isNotEmpty()) {
+            request.addRequestHeader("Referer", refererURL)
+        }
+
+        request.setDestinationInExternalPublicDir(download.destinationDirectory, fileName)
+
+        val downloadID = androidDownloadManager.enqueue(request)
+        queuedDownloads[downloadID] = downloadListener
+
+        if (!isSubscribedReceiver)
+            registerBroadcastReceiver(applicationContext)
+
+        return downloadID
+    }
+
+    /**
+     * Use for register a listener for scheduled download.
+     * in case you did not provide one on [download] function.
+     * @param downloadID the id that reference to the schedule download.
+     * @param listener a listener to be notify when the download is completed.
+     */
+    @Suppress("unused")
+    fun registerListener(downloadID: Long, listener: OnDownloadFinish) {
+        if (downloadID in queuedDownloads)
+            queuedDownloads[downloadID] = listener
+        else
+            throw IllegalArgumentException("Unable to find $downloadID on the queued downloads")
+    }
+
+    /**
+     * Remove all the listeners.
+     */
+    fun unregisterListeners() {
+        applicationContext.unregisterReceiver(onDownloadComplete)
+        isSubscribedReceiver = false
+        queuedDownloads.clear()
+    }
```

Comment:
Your listener setup is very fine-grained and I wonder if we can do it simpler. The apps we have are not really interested in subscribing to specific downloads. For example Focus shows a `Snackbar` whenever a download is completed. For that some simple callback would be enough:

```
onDownloadCompleted(download: Download, ... additional meta data if needed)
```

This callback (if registered) could get invoked for all downloads and I think I'd want to register that on the `Feature` when initializing it.

Does this make sense?